Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs

```diff
@@ -197,6 +197,50 @@ ResponderDispatchService responderDispatch
         _payloadsToSend.Enqueue(payload);
     }
 
+    /// <summary>
+    /// Uses a given gateway session to connect to the gateway with.
+    /// </summary>
+    /// <param name="sessionInformation">The gateway session to connect with.</param>
+    /// <returns>A result that may or not have succeeded.</returns>
+    public Result UseGatewaySession(GatewaySessionInformation sessionInformation)
+    {
+        if (_connectionStatus is not GatewayConnectionStatus.Offline)
+        {
+            return new InvalidOperationError("Cannot use a gateway session while the client is connected.");
+        }
+
+        _sessionID = sessionInformation.SessionID;
+        _lastSequenceNumber = sessionInformation.SequenceNumber;
+        _isSessionResumable = true;
+
+        return Result.FromSuccess();
+    }
+
+    /// <summary>
+    /// Returns information about the current gateweay session.
+    /// </summary>
+    /// <remarks>
+    /// This method should only be called after the client has disconnected.
+    /// If the client reconnects, or is still running, the session ID and sequence number may be invalid.
+    /// </remarks>
+    /// <returns>A result containing the current gateway session information.</returns>
+    public Result<GatewaySessionInformation> GetGatewaySessionInformation()
+    {
+        if (string.IsNullOrEmpty(_sessionID) || _lastSequenceNumber is 0)
+        {
+            return Result<GatewaySessionInformation>.FromError
+            (
+                new InvalidOperationException("No session information is available.")
```

Comment:
A `NotFoundError` might be more appropriate. Also, you should be able to just return the result directly.

Reply:
The thought process was that the gateway client always preserves its last known state with the gateway so that it can be extracted by the user in a safe manner (e.g. to be resumed after a startup)

Reply:
So, if this error is returned, the only possible state is that the client was never started, and thusly this method is being used incorrectly. So perhaps an exception, because of developer error.

Reply:
I assume you wanted to reply to the other comment :P but, as I said, the points where the session ID is nulled out  are points where we know for a fact that the session is *not* resumable, and as such there's no point in keeping the information around. It's more likely to cause issues than anything.